source/nodejs/adaptivecards-templating/src/expression-parser.ts

```diff
@@ -0,0 +1,772 @@
+    "*",
+    "-",
+    "+",
+    "==",
+    "!=",
+    "<",
+    "<=",
+    ">",
+    ">="
+];
+
+const literals: Array<TokenType> = [
+    "identifier",
+    "string",
+    "number",
+    "boolean"
+];
+
+interface TokenizerRule {
+    tokenType: TokenType;
+    regEx: RegExp;
+}
+
+interface Token {
+    type: TokenType;
+    value: string;
+    originalPosition: number;
+}
+
+class Tokenizer {
+    static rules: Array<TokenizerRule> = [];
+
+    static init() {
+        Tokenizer.rules.push(
+            { tokenType: undefined, regEx: /^\s/ },
+            { tokenType: "{", regEx: /^{/ },
+            { tokenType: "?#", regEx: /^\?#/ },
+            { tokenType: "}", regEx: /^}/ },
+            { tokenType: "[", regEx: /^\[/ },
+            { tokenType: "]", regEx: /^\]/ },
+            { tokenType: "(", regEx: /^\(/ },
+            { tokenType: ")", regEx: /^\)/ },
+            { tokenType: "boolean", regEx: /^true|^false/ },
+            { tokenType: "identifier", regEx: /^[$a-z_]+/i },
+            { tokenType: ".", regEx: /^\./ },
+            { tokenType: ",", regEx: /^,/ },
+            { tokenType: "+", regEx: /^\+/ },
+            { tokenType: "-", regEx: /^-/ },
+            { tokenType: "*", regEx: /^\*/ },
+            { tokenType: "/", regEx: /^\// },
+            { tokenType: "==", regEx: /^==/ },
+            { tokenType: "!=", regEx: /^!=/ },
+            { tokenType: "<=", regEx: /^<=/ },
+            { tokenType: "<", regEx: /^</ },
+            { tokenType: ">=", regEx: /^>=/ },
+            { tokenType: ">", regEx: /^>/ },
+            { tokenType: "string", regEx: /^"([^"]*)"/ },
+            { tokenType: "string", regEx: /^'([^']*)'/ },
+            { tokenType: "number", regEx: /^\d*\.?\d+/ }
+        )
+    }
+
+    static parse(expression: string): Token[] {
+        let result: Token[] = [];
+        let i = 0;
+
+        while (i < expression.length) {
+            let subExpression = expression.substring(i);
+            let matchFound = false;
+
+            for (let rule of Tokenizer.rules) {
+                let matches = rule.regEx.exec(subExpression);
+
+                if (matches) {
+                    if (matches.length > 2) {
+                        throw new Error("A tokenizer rule matched more than one group.");
+                    }
+
+                    if (rule.tokenType != undefined) {
+                        result.push(
+                            {
+                                type: rule.tokenType,
+                                value: matches[matches.length == 1 ? 0 : 1],
+                                originalPosition: i
+                            }
+                        )
+                    }
+
+                    i += matches[0].length;
+
+                    matchFound = true;
+
+                    break;
+                }
+            }
+
+            if (!matchFound) {
+                throw new Error("Unexpected character " + subExpression[0] + " at position " + i);
```

Comment:
>Error [](start = 26, length = 5)

Nit: maybe we should end up using our own custom error class here if one doesnt already exist instead of generic Error

Reply:
Derived error classes are nonstandard in JS/TS. There is no way to catch a specific error type as far as I know, one has to write something like this (assuming RangeError extends Error):
```typescript
try {
    throw new RangeError();
}
catch (e) {
    if(e instanceof RangeError){
        console.log('out of range');
    }
}
```
So essentially you always throw a generic Error anyway.

Reply:
Hmm TS isnt unique in that error handling pattern -- other languages also enable custom error handling code that looks like that.. So that doesnt seem justifcation to not use custom errors in the code base; Infact preferring custom error types does seem to be a best practice of sorts and not that unheard of in TS/JS -- found this --https://blog.joefallon.net/2018/09/typescript-try-catch-finally-and-custom-errors/ and it lists precisely this as a best practice - "Best Practice – Be Specific with Errors
Often you will want to throw or return errors that are more specific than the generic Error object. That is easy to do by extending the standard error object. The goal with having a more specific error is to maximize the chance that the error can be handled locally and allow code execution to continue. Here is how you catch the specific errors."  I think we should revisit this pattern in our code base.. can probably sync up a bit offline as well as something we can work on together later?

---
In reply to: [297901240](https://github.com/microsoft/AdaptiveCards/pull/2995#discussion_r297901240) [](ancestors = 297901240)